File: lc/T101-150/T106.cc
#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution
{
public:
    int pos = 0;
    map<int, int> m;

    TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder)
    {

        int id = 0;
        for (auto it : inorder)
        {
            m[it] = id++;
        }
        int now = postorder.size() - 1;
        return travel(inorder, postorder, now, 0, inorder.size() - 1);
    }
    TreeNode *travel(vector<int> &inorder, vector<int> &postorder, int &now, int l, int r)
    {
        cout << " now : " << now << " l: " << l << " r: " << r << "id " << endl;
        if (l > r || now < 0)
            return nullptr;
        TreeNode *root = new TreeNode(postorder[now]);
        auto id = m[postorder[now]];

        now--;
        root->right = travel(inorder, postorder, now, id + 1, r);
        root->left = travel(inorder, postorder, now, l, id - 1);
        return root;
    }
};

void dfs(TreeNode *t)
{
    if (t == nullptr)
        return;
    cout << t->val << " ";
    dfs(t->left);
    dfs(t->right);
}
int main()
{
    Solution s;

    vector<int> v1{9, 3, 15, 20, 7}, v2{9, 15, 7, 20, 3};
    auto l = s.buildTree(v1, v2);
    dfs(l);
}

File: lc/T101-150/t_106.java


package leetcoder;

import java.util.*;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.LinkedBlockingQueue;
//
//
class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if (inorder.length == 0) return null;
        return get(0, inorder.length - 1,0,postorder.length-1,inorder, postorder);
    }

    public TreeNode get(int ll, int lr,int rl,int rr ,int[] inorder, int[] postorder) {
        if( ll > lr  ) return null;

        TreeNode t = new TreeNode(postorder[rr]);
        int index = -1 ;
        for(int i = ll;i<=lr;i++){
            if(inorder[i] == postorder[rr]){
                index = i;
                break;
            }
        }
        if(index==-1) return null;

        t.left = get(ll,index-1,rl,rl+ (index -ll -1),inorder,postorder);
        t.right = get(index+1,lr,rl + (index-ll) ,rr-1,inorder,postorder);
        return t;

    }
}

class Main {
    public static void preTravel(TreeNode t) {
        if (t == null) return;
        System.out.print(t.val);
        preTravel(t.left);
        preTravel(t.right);
    }
    public static void main(String[] arg) {



        int [] inorder = {9,3,15,20,7};
        int [] postorder = {9,15,7,20,3};
        TreeNode t = new leetcoder.Solution().buildTree(inorder,postorder);
        preTravel(t);
        System.out.println();

    }
}